File: plugins/sulis-execution/sdk/typescript/src/resources/lifecycle.ts
import { kwargsToParams, resultPayload } from '../helpers.js';
import type {
  AsyncSubprocessTransport,
  SubprocessTransport,
  TransportConfig,
} from '../transport.js';
import type { LifecycleCompleteResult } from '../types.js';

/**
 * Lifecycle resource — atomically complete a Work Package's lifecycle.
 *
 * Underlying CLI: `wpx-step12 wrap`. The SDK wraps that under cleaner
 * names (`lifecycle.complete` vs the CLI's `step12 wrap`).
 */

const BINARY = 'wpx-step12';
const SUBCOMMAND = 'wrap'; // SDK method is `complete`; CLI keeps `wrap`

function common(config: TransportConfig): Record<string, unknown> {
  return { project: config.project, repo_root: config.repoRoot };
}

export interface LifecycleCompleteParams {
  wp: string;
  branch: string;
  pipeline_result: string;
  pre_squash_sha?: string;
  worktree_path?: string;
  post_deploy_verification?: string;
}

export class LifecycleResource {
  constructor(
    private readonly transport: SubprocessTransport,
    private readonly config: TransportConfig,
  ) {}

  /**
   * Atomically complete a Work Package's lifecycle.
   *
   * Chains three operations fail-fast: append evidence to the WP file,
   * flip INDEX status to `done` (expected: in_progress → done), remove
   * the worktree. If any fails, returns details of what succeeded vs
   * failed.
   */
  complete(params: LifecycleCompleteParams): LifecycleCompleteResult {
    return resultPayload(
      this.transport.invoke(BINARY, SUBCOMMAND, {
        ...common(this.config),
        ...kwargsToParams(params),
      }),
    ) as unknown as LifecycleCompleteResult;
  }
}

export class AsyncLifecycleResource {
  constructor(
    private readonly transport: AsyncSubprocessTransport,
    private readonly config: TransportConfig,
  ) {}

  async complete(
    params: LifecycleCompleteParams,
  ): Promise<LifecycleCompleteResult> {
    return resultPayload(
      await this.transport.invoke(BINARY, SUBCOMMAND, {
        ...common(this.config),
        ...kwargsToParams(params),
      }),
    ) as unknown as LifecycleCompleteResult;
  }
}

File: plugins/sulis-execution/sdk/python/sulis_execution/resources/lifecycle.py
"""Lifecycle resource — atomically complete a Work Package's lifecycle.

After the pipeline succeeds, `complete()` atomically finalises a WP:
appends evidence to the WP file, flips INDEX status to `done`, removes
the worktree. All three or none — fail-fast.

Underlying CLI: `wpx-step12 wrap`. The SDK wraps that under cleaner names.
"""
from __future__ import annotations

from typing import Optional

from sulis_execution._helpers import _kwargs_to_params, _result_payload
from sulis_execution.transport import (
    AsyncSubprocessTransport,
    SubprocessTransport,
    TransportConfig,
)
from sulis_execution.types import LifecycleCompleteResult

BINARY = "wpx-step12"
SUBCOMMAND = "wrap"  # underlying CLI subcommand; SDK method is `complete`


def _common(config: TransportConfig) -> dict:
    return {"project": config.project, "repo_root": str(config.repo_root)}


class LifecycleResource:
    def __init__(
        self, transport: SubprocessTransport, config: TransportConfig
    ) -> None:
        self._transport = transport
        self._config = config

    def complete(
        self,
        *,
        wp: str,
        branch: str,
        pipeline_result: str,
        pre_squash_sha: Optional[str] = None,
        worktree_path: Optional[str] = None,
        post_deploy_verification: Optional[str] = None,
    ) -> LifecycleCompleteResult:
        """Atomically complete a Work Package's lifecycle.

        Chains three operations fail-fast:
        1. work_package.append_evidence
        2. index.flip_status (expected: in_progress → done)
        3. worktree.remove

        If any fails, returns details of what succeeded vs failed.
        """
        params = _kwargs_to_params({
            **_common(self._config),
            "wp": wp, "branch": branch, "pipeline_result": pipeline_result,
            "pre_squash_sha": pre_squash_sha, "worktree_path": worktree_path,
            "post_deploy_verification": post_deploy_verification,
        })
        envelope = self._transport.invoke(BINARY, SUBCOMMAND, params)
        return LifecycleCompleteResult.model_validate(_result_payload(envelope))


class AsyncLifecycleResource:
    def __init__(
        self, transport: AsyncSubprocessTransport, config: TransportConfig
    ) -> None:
        self._transport = transport
        self._config = config

    async def complete(
        self,
        *,
        wp: str,
        branch: str,
        pipeline_result: str,
        pre_squash_sha: Optional[str] = None,
        worktree_path: Optional[str] = None,
        post_deploy_verification: Optional[str] = None,
    ) -> LifecycleCompleteResult:
        params = _kwargs_to_params({
            **_common(self._config),
            "wp": wp, "branch": branch, "pipeline_result": pipeline_result,
            "pre_squash_sha": pre_squash_sha, "worktree_path": worktree_path,
            "post_deploy_verification": post_deploy_verification,
        })
        envelope = await self._transport.invoke(BINARY, SUBCOMMAND, params)
        return LifecycleCompleteResult.model_validate(_result_payload(envelope))
